### DISCUSSION TITLE
Camera setting conversion for MeshLab 2021.10

### DISCUSSION THREAD

**Discussion by janhuang6** [cnr-isti-vclab/meshlab#1169] [General]
MeshLab 2021.10 camera setting is in this format:

TranslationVector="1.36674 -3.59469 -40.0356 1" RotationMatrix="1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 " CameraType="0" PixelSizeMm="0.0369161 0.0369161" CenterPx="484 387" ViewportPx="969 774" LensDistortion="0 0" BinaryData="0" FocalMm="24.74498"
TrackScale="0.068330936" FarPlane="6.0881977" NearPlane="0.30310887"

In my project, camera setting is given in this format:

#focal length
22977.3 22917.8
#resolution
3482 5218
#principal point
1741 2609
#distortion coeffs
0.641043 -22.3842 0 0
#camera transformation matrix 1
Rotation Euler X Y Z (degrees): 0.525084 -48.0539 0.176361
Translation : -119.556	3.04448	103.007
MATRIX :
0.668428	-0.0098941	-0.743711	-119.556
0.00205749	0.999932	-0.0114536	3.04448
0.743774	0.00612572	0.668403	103.007
0	0	0	1

Now I am having difficulty to convert from my camera setting to MeshLab's format.
What is MeshLab's PixelSizeMm, LensDistortion, BinaryData, FocalMm, TrackScale, FarPlane, NearPlane? How to convert from my format to MeshLab's for those variables? 

Is RotationMatrix always fixed as "1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 ", and BinaryData always "0" in MeshLab, or at least that is the case for my application?

Thank you for any input!

**Comment by Craig-SA** [comment]
I am developing my own software for a research dissertation. I have camera position coordinates in x,y,z,w format. How do I replicate these in Meshlab?

I want to compare a cross-section taken from my software to one taken in Meshlab. To do so I need to replicate the camera position in Meshab.

I know how to copy camera settings from Meshlab, and how to update the camera settings back to Meshlab. I just don't know what the fields in the camera settings mean (similar to question above). 

Any help appreciated. Thanks.

**Comment by jmespadero** [comment]
> Is RotationMatrix always fixed as "1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 ", and BinaryData always "0" in MeshLab?

No. RotationMatrix is initialized to "1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 " when meshlab is started, but the value will change if you orbit the camera clicking and dragging in the scene.

>  Is BinaryData always "0" ?

Currently, yes. This is a remain of old code that saved camera data in binary format instead of XML. 

> What is MeshLab's PixelSizeMm, LensDistortion, FocalMm, TrackScale, FarPlane, NearPlane? How to convert from my format to MeshLab's for those variables?

* PixelSizeMm: Declared size of one pixel in your display. 

* LensDistortion, 

* **TrackScale** Scale applied to the camera, so similar to apply to all geometry of the scene. You can change TrackScale using the mouse wheel to see the effect.
![TrackScale change](https://github.com/cnr-isti-vclab/meshlab/assets/6451497/00776c52-e786-45dc-8551-1b61efb95343)

* **FocalMm**: Focal distance (focal length) of the camera (measured in mm). A longer focal distance generally results in a narrower field of view and higher magnification, while a shorter focal length provides a wider field of view and lower magnification. You can change its value using the mouse wheel while pressing Shift to see the effect.
![FocalMm change](https://github.com/cnr-isti-vclab/meshlab/assets/6451497/38607a8d-cdf5-4683-8372-8b156a3b5d07)

* **FarPlane**, **NearPlane** : Those are the values used for Z-clipping in openGL, so you can hide geometry which is near to the camera. You can change NearPlane using the mouse wheel while pressing Ctrl to see the effect. If unsure, let NearPlane=0.1 anf FarPlane=999999.0

![Z-Clip using NearPlane](https://github.com/cnr-isti-vclab/meshlab/assets/6451497/8cf94a4a-3a26-4334-b816-3ad109ced240)

Your project uses two values to "Focal Length", perhaps different values per horizontal and vertical axis?. I would use the average value.

The "Rotation Euler" contains same info than RotationMatrix, so can be ignored.


## Implementation details:

The information of the ViewState will be written as a XML string to the clipboard. You can paste this screen into notepad or other text editor. The XML string has always this format 

```xml
<!DOCTYPE ViewState>
<project>
 <VCGCamera 
        BinaryData="0" 
        TranslationVector="-1312.22 -1719.2 -2217.1 1" 
        RotationMatrix="0.690565 -0.0230493 -0.722903 0 -0.491698 0.71804 -0.492597 0 0.530427 0.69562 0.48452 0 0 0 0 1 "
        PixelSizeMm="0.0369161 0.0369161" 
        ViewportPx="1479 921" 
        LensDistortion="0 0" 
        CameraType="0" 
        FocalMm="29.444607509282928" 
        CenterPx="739 460" 
/>
 <ViewSettings 
        FarPlane="8.5907297" 
        NearPlane="0.30310887" 
        TrackScale="0.001281216" 
/>
</project>

```
The code that writes this XML string is [GLArea::viewToText()](https://github.com/cnr-isti-vclab/meshlab/blob/bd88167db9839109487f401be50991c4bc990d27/src/meshlab/glarea.cpp#L2192), and the code that read the XML string and apply to the trackball is [GLArea::loadViewFromViewStateFile(const QDomDocument &doc)](https://github.com/cnr-isti-vclab/meshlab/blob/bd88167db9839109487f401be50991c4bc990d27/src/meshlab/glarea.cpp#L2153).

As you see, the XML has two sections:
* VCGCamera section, which parsing is delegated to [ReadShotFromQDomNode()](https://github.com/cnr-isti-vclab/vcglib/blob/0ca039b6a6058c7170b3f19971848af338d06ca7/wrap/qt/shot_qt.h#L14), and then applied to current glarea calling to [loadShot(QPair<Shotm, float> (shot,trackball.track.sca))](https://github.com/cnr-isti-vclab/meshlab/blob/bd88167db9839109487f401be50991c4bc990d27/src/meshlab/glarea.cpp#L2281), which will change scale, orientation and position of the camera in [shot2Track()](https://github.com/cnr-isti-vclab/meshlab/blob/bd88167db9839109487f401be50991c4bc990d27/src/meshlab/glarea.h#L667). 
* ViewSettings section, which is applied directly to member values of current glarea,


  [1]: https://i.stack.imgur.com/aZxMA.png
